src/composables/useLoop/index.ts

```diff
@@ -0,0 +1,64 @@
+import type { EventDispatcher, Raycaster, WebGLRenderer } from 'three'
+import { unref } from 'vue'
+import type { TresCamera, TresScene } from '../../types'
+import { useTresContext } from '../useTresContextProvider'
+import type { LoopCallback, LoopCallbackParams } from '../../core/loop'
+import type { SizesType } from '../useSizes'
+
+export interface LoopParams extends LoopCallbackParams {
+  camera: TresCamera
+  scene: TresScene
+  renderer: WebGLRenderer
+  invalidate: () => void
+  advance: () => void
+  raycaster: Raycaster
+  controls: (EventDispatcher & { enabled: boolean })
+  sizes: SizesType
+}
+
+export type Fn = (params: LoopParams) => void
+
+export function useLoop() {
+  const {
+    camera,
+    scene,
+    renderer,
+    loop,
+    raycaster,
+    controls,
+    invalidate,
+    advance,
+  } = useTresContext()
+
+  function onBeforeRender(cb: Fn, index = 0) {
+    const wrappedCallback = (params: LoopCallbackParams) => {
+      cb({ ...params, camera: unref(camera) as TresCamera, scene: unref(scene), renderer: unref(renderer), raycaster: unref(raycaster), controls: unref(controls), invalidate, advance })
+    }
+    loop.register(wrappedCallback as LoopCallback, 'before', index)
+  }
+
+  function render(cb: Fn) {
+    const wrappedCallback = (params: LoopCallbackParams) => {
+      cb({ ...params, camera: unref(camera) as TresCamera, scene: unref(scene), renderer: unref(renderer), raycaster: unref(raycaster), controls: unref(controls), invalidate, advance })
+    }
+    loop.register(wrappedCallback as LoopCallback, 'render')
+  }
+
+  function onAfterRender(cb: Fn, index = 0) {
+    const wrappedCallback = (params: LoopCallbackParams) => {
+      cb({ ...params, camera: unref(camera) as TresCamera, scene: unref(scene), renderer: unref(renderer), raycaster: unref(raycaster), controls: unref(controls), invalidate, advance })
+    }
+    loop.register(wrappedCallback as LoopCallback, 'after', index)
+  }
```

Comment:
```suggestion
  const createWrappedCallback = (cb: Fn) => (params: LoopCallbackParams) => 
    cb({
      ...params,
      camera: unref(camera) as TresCamera,
      scene: unref(scene),
      renderer: unref(renderer),
      raycaster: unref(raycaster),
      controls: unref(controls),
      invalidate,
      advance
    })

  const onBeforeRender = (cb: Fn, index = 0) =>
    loop.register(createWrappedCallback(cb), 'before', index)

  const render = (cb: Fn) =>
    loop.register(createWrappedCallback(cb), 'render')

  const onAfterRender = (cb: Fn, index = 0) => 
    loop.register(createWrappedCallback(cb), 'after', index)
```

Reply:
@Tinoooo this is solved with the approach Andre provided